Use optparse module

avdv opened the pull request:
Fixes #74 and fixes #83.

Please have a look.. :smiley: 

Another gimmick: optparse can generate completion code for bash and zsh:

```bash
$ colorls --'*'-completion-bash=--a
--all
--almost-all
```

avdv commented on lib/colorls/core.rb:
```diff
@@ -1,7 +1,7 @@
   class Core
     def initialize(input=nil, all: false, report: false, sort: false, show: false,
-      one_per_line: false, long: false, almost_all: false, tree: false, help: false, colors: [])
+      one_per_line: false, long: false, almost_all: false, tree: false, colors: [])
```
"better" is just a matter of taste, I guess.

But usually you want to avoid any tedious or repetitious (maintenance) work; so I wanted to keep this DRY.

Also I did not understand the message of the helplog, what the "command" part was referring to... something about grep and the README? :confused:

BTW, here's the output after applying this change:

```
Usage:  colorls [OPTION]... [FILE]...

    -a, --all                        do not ignore entries starting with .
    -A, --almost-all                 do not list . and ..
    -l, --long                       use a long listing format
    -t, --tree                       shows tree view of the directory
    -r, --report                     show brief report
    -1                               list one file per line
    -d, --dirs                       show only directories
    -f, --files                      show only files

sorting options:

    -sMODE                           sort (f)iles or (d)irectories first
        --sort-dirs, --group-directories-first
                                     sort directories first
        --sort-files                 sort files first

general options:

    -h, --help                       Prints this help
        --version                    Show version
```

athityakumar replied:
> Also I did not understand the message of the helplog, what the "command" part was referring to... something about grep and the README? 😕

Basically, all misc use-cases. Like, `colorls <filename>`, `colorls *` and `colorls <directory> -sd | grep <search>`

athityakumar commented on lib/colorls/flags.rb:
```diff
@@ -34,6 +36,52 @@ def process
+      end
+      options.on('--sort-dirs', '--group-directories-first', 'sort directories first') { @opts[:sort] = :dirs }
+      options.on('--sort-files', 'sort files first')                                   { @opts[:sort] = :files }
+    end
```
Please add `-sd` as short-hand of `--sort-dirs`, and similarly `-sf` for `--sort-files`

avdv replied:
I already did. But since optparse does not support multi character short options (AFAICS), I had to model that using a required parameter to the newly introduced short option `-s` (similar to what @GladOSkar suggested [here](https://github.com/athityakumar/colorls/issues/74#issuecomment-314454679)

`-sMODE                           sort (f)iles or (d)irectories first`

MODE could be 'f' or 'd' which mimics the behavior from before. Any other character raises an error.

athityakumar replied:
Peace. Let's leave this as it is.

avdv replied:
Maybe the help text could be improved here... something like

`-sTYPE   show TYPE first, -sd = directories first, -sf = files first`

athityakumar replied:
`-sTYPE show TYPE first, -sd = directories first, -sf = files first` seems better. 😄 

1. I'm not sure of how `optparse` implements TAB auto-complete, but does `-sd` & `-sf` auto-complete?
2. Maybe we should rename `-sf`, `-sd` to `-F`, `-D`? But this would break these tags for existing users, and I'm not quite with this idea.

avdv replied:
1. no, optparse only completes the Option itself: `-s`. I'm looking into somehow overriding the behavior of the OptionParser to permit these two options... but I have to read up a bit more on the Ruby syntax and optparse docs.

2. I would like to get rid of these options (see #103), but I would not want to break it in this PR just yet

avdv replied:
I have changed the help text accordingly.

Alas, concerning adapting the optparse logic I was not successful. Maybe it is best to leave it this way until the refactoring of the options...

athityakumar replied:
No issues. Let's deprecate `-sd` / `-sf` and also introduce `--sd` / `--sf` for optparse auto-completion. Also, can you add the misc use cases like `colorls *`, and `colorls | grep`?

athityakumar commented:
@avdv - Let's deprecate `-sd` / `-sf` and also introduce `--sd` / `--sf` for optparse auto-completion. Would you like to include that in this PR, or reserve them for later?

Also, can you add the misc use cases like `colorls *`, and `colorls | grep`?

athityakumar commented:
@avdv - Please rebase with master, and add `-gs` / `--git-status` flag to optparse please. 😄

avdv commented:
Ok, I will. Just don't hold your breath... I am quite busy currently. 😉

athityakumar commented:
Haha, sure. 😄

athityakumar commented:
Ping @avdv - Just wanted to remind you about this PR.

avdv commented:
Hi @athityakumar, sorry for the long delay! I have been pretty much loaded with work, and then needed a vacation... :wink:

I hope that I will find some time finishing this PR in the next few days! Thanks for your patience.

avdv commented:
Hi, @athityakumar, I have rebased this branch and updated the PR to include the changes you requested:

```
$ colorls -h
Usage:  colorls [OPTION]... [FILE]...

    -a, --all                        do not ignore entries starting with .
    -A, --almost-all                 do not list . and ..
    -l, --long                       use a long listing format
    -t, --tree                       shows tree view of the directory
    -r, --report                     show brief report
    -1                               list one file per line
    -d, --dirs                       show only directories
    -f, --files                      show only files
        --gs, --git-status           show git status for each file

sorting options:

        --sd, --sort-dirs, --group-directories-first
                                     sort directories first
        --sf, --sort-files           sort files first

general options:

        --light                      use light color scheme
        --dark                       use dark color scheme

    -h, --help                       Prints this help
        --version                    Show version

examples:

  * show the given file:

    colorls README.md

  * show matching files and list matching directories:

    colorls *

  * filter output by a regular expression:

    colorls | grep PATTERN

```

avdv commented:
> Maybe the help menu should also explicitly state that the flags can be clubbed together. Add this, and I'll merge this soon

Huh? What happened, you already merged it... 😄 Thanks!

I could still add an example which demonstrates clubbing short flags togehter?

athityakumar commented:
@avdv - Ah, sure. Feel free to open a new PR. 😄
